### DISCUSSION TITLE
Add an option to choose between 16 or 24 bit processing and output

### DISCUSSION THREAD

**Discussion by fuslwusl** [music-assistant/support#1599] [Feature requests and ideas]
Most AV receivers and renderer support 24 bit audio as Flac or WAV stream.
It would be great to have an option to choose between 16 and 24 bit output and processing to avoid loss in dynamic resolution when volume normalization is active.

@marcelveldt gave some very interesting insights in how audio is handled inside music assistant #1501 :

> All tracks are processed as raw pcm by Music Assistant internally. So everything you play will be first converted to just raw pcm in the sample rate and bit depth of the origin (unless flow mode is enabled) and the gain adjustment is done while extracting the raw source media so the pcm chunks we get into the streaming engine have the gain adjustment applied, that way there should be enough headroom withit the 16 or 24 bits sample rate. ... I settled on applying the correction gain while extracting the raw audio from its base codec. ...
> The final part in the chain is that MA sends either the raw pcm data to the player or encodes it into a codec of the user's choice, by default this is FLAC due to it being lossless while keeping a sane file size ... 

Currently there is an setting for normalization level (EBU) and output codec but not for the output bit depth the device is able to use and the volume normalization should be processed.
So when i play a 16 bit 44.1 kHz lossless FLAC track from Deezer it will lose dynamic resolution when normalization is active because the volume is reduced (in worst case 17 dB with default settings) in 16 bit resolution (because the input track was 16 bit).

For higher quality audio devices like av amps or speakers it will be great to have the full 24 bit resolution used for processing. So a 16 bit file will not lose dynamic resolution when played back at 24 bit with volume normalization.

**Comment by marcelveldt** [comment]
We can actually just send 24 bits by default for players that support it, that would even be better. On high performance systems we can also use 64 bits floating point for all audio processing which is actually even better but a tiny bit more resource hungry and only go back to 24 bits in the final stage when we send audio to the player.

The idea is good, I'll think about it for a while how to implement it in a way that the default is good.

**Reply by fuslwusl** [reply]
In my opinion 24 bit for audio is enough in nearly all cases but it would be great to preserve the full dynamic resolution of 16 bit source tracks when volume normalization is active.

I checked the documentation of some devices:

* Denon https://manuals.denon.com/HEOSAVR/ALL/DE/OKNRSYyqodrnii.php
* Sony https://helpguide.sony.net/ha/strdn10585/v1/en/contents/TP0000378902.html
* Yamaha https://de.yamaha.com/de/products/audio_visual/accessories/wxad-10/specs.html
* Chromecast https://support.google.com/chromecast/answer/6279377?hl=en-CA

Nearly all av receivers and amps support FLAC 24 bit and WAV 24 bit.
Chromecast Audio, Google Home and Google Home Mini support FLAC with 24 bit, too.

The max. supported bit depth depends on the codec used:
24 bit: WAV or FLAC
16 bit: MP3

For all av receiver owners - check out UPnP in musicassistant in combination with your receivers because volume control and everything else works so great with musicassistant! And you will not get any better replacement for the Denon HEOS crap!!!

**Reply by OzGav** [reply]
I believe this is implemented. MA maintains the quality of the source if it can given any constraints of the player. The awesome AUDIO PIPELINE view shows you the path it takes

**Reply by marcelveldt** [reply]
Correct, this is implemented.

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
